### PR TITLE
Directive improvements

### PR DESCRIPTION
**Pull request by lmreis89**
1. Headers, cookies, and body written into the context by pac4j get put into the HttpResponse
2. Add content type to HttpAdapter
3. Support form encoding in the directive
4. Remove sbt formatting plugins

### CODE REVIEW

File: src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala [hunk 1]
```diff
@@ -2,28 +2,67 @@ package com.stackstate.pac4j
...
+import scala.concurrent.{ExecutionContext, Future}
+import scala.collection.immutable
 
 object AkkaHttpSecurity {
   type AkkaHttpSecurityLogic = SecurityLogic[Future[RouteResult], AkkaHttpWebContext]
 
   def authorize(authorizer: Authorizer[CommonProfile])(request: AuthenticatedRequest): Directive0 =
     akkaHttpAuthorize(authorizer.isAuthorized(request.webContext, request.profiles.asJava))
+
+  private def mapCookie(cookie: Cookie): HttpCookie = {
```
Comment (lmreis89) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala]: Would change it to 'toAkkaCookie' or something of sorts.

File: src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala [hunk 2]
```diff
@@ -43,16 +82,37 @@ class AkkaHttpSecurity[P <: CommonProfile](config: Config) {
...
       override def tapply(innerRoute: Tuple1[AuthenticatedRequest] => Route): Route = { ctx =>
-        val context = AkkaHttpWebContext(ctx.request)
-        securityLogic.perform(context, config, (context: AkkaHttpWebContext, profiles: util.Collection[CommonProfile], parameters: AnyRef) => {
-          val authenticatedRequest = AuthenticatedRequest(context, profiles.asScala.toList)
-          innerRoute(Tuple1(authenticatedRequest))(ctx)
-          // TODO: Add response headers from the AkkaHttpWebContext and put them into the response
-        }, actionAdapter, clients, "", "", multiProfile)
+        import ctx.materializer
+
+        Unmarshal(ctx.request.entity).to[StrictForm].fast.flatMap { form =>
```
Comment (lmreis89) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala]: Need to add some comments here

File: src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala [hunk 3]
```diff
@@ -1,21 +1,24 @@
...
-case class AkkaHttpWebContext(request: HttpRequest) extends WebContext {
+case class AkkaHttpWebContext(request: HttpRequest, formFields: Seq[(String, String)] = Seq.empty) extends WebContext {
   import com.stackstate.pac4j.AkkaHttpWebContext._
 
   private var changes = ResponseChanges.empty
   private lazy val requestCookies = request.cookies.map { akkaCookie =>
     new Cookie(akkaCookie.name, akkaCookie.value)
   }.asJavaCollection
 
+  private lazy val requestParameters = formFields.toMap ++ request.getUri().query().toMap.asScala
```
Comment (lmreis89) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala]: Some comments here would be great

File: src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala [hunk 4]
```diff
@@ -45,13 +48,16 @@ case class AkkaHttpWebContext(request: HttpRequest) extends WebContext {
   }
 
   override def setResponseContentType(contentType: String): Unit = {
-    changes = changes.copy(contentType = contentType)
+    ContentType.parse(contentType) match {
+      case Right(ct) =>
+        changes = changes.copy(contentType = ct)
+      case Left(_) =>
+        throw new IllegalArgumentException("Invalid response content type " + contentType)
```
Comment (lmreis89) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala]: I think there's an Akka exception for this already so we could reuse that

File: src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala [hunk 5]
```diff
@@ -45,13 +51,16 @@ case class AkkaHttpWebContext(request: HttpRequest) extends WebContext {
   }
 
   override def setResponseContentType(contentType: String): Unit = {
-    changes = changes.copy(contentType = contentType)
+    ContentType.parse(contentType) match {
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala]: I like this parsing of the contentType in the setter already, lets do this for cookies aswell. We can make this into a separate MR (or put it on top of this one)

File: src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala [hunk 6]
```diff
@@ -94,22 +103,30 @@ case class AkkaHttpWebContext(request: HttpRequest) extends WebContext {
...
 }
 
 object AkkaHttpWebContext {
 
   //This class is where all the HTTP response changes are stored so that they can later be applied to an HTTP Request
-  case class ResponseChanges private(
+  case class ResponseChanges private (
     headers: List[(String, String)],
-    contentType: String,
+    contentType: ContentType,
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala]: I would make this 'Option' to signify the situation that no ContentType was set

File: src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala [hunk 7]
```diff
@@ -94,22 +103,30 @@ case class AkkaHttpWebContext(request: HttpRequest) extends WebContext {
...
     content: String,
     cookies: List[Cookie],
     attributes: Map[String, AnyRef])
 
   object ResponseChanges {
     def empty: ResponseChanges = {
-      ResponseChanges(List.empty, "", "", List.empty, Map.empty)
+      ResponseChanges(List.empty, ContentTypes.`text/plain(UTF-8)`, "", List.empty, Map.empty)
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpWebContext.scala]: If ContentType would be Optional, we wouldn't have to invent a contenttype here

File: src/test/scala/com/stackstate/pac4j/AkkaHttpActionAdapterTest.scala [hunk 8]
```diff
@@ -28,5 +28,9 @@ class AkkaHttpActionAdapterTest extends WordSpecLike with Matchers with ScalaFut
     "convert 204 to NoContent" in {
       AkkaHttpActionAdapter.adapt(204, dummyContext).futureValue.response shouldEqual HttpResponse(NoContent)
     }
+    "convert 200 to OK with content set from the context" in {
```
Comment (craffit) [inline src/test/scala/com/stackstate/pac4j/AkkaHttpActionAdapterTest.scala]: I'm missing a test here which actually checks for the set content type

File: src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala [hunk 9]
```diff
@@ -43,16 +82,43 @@ class AkkaHttpSecurity[P <: CommonProfile](config: Config) {
    * this does not apply any authorization ofr filtering.
    */
   def withAuthentication(
-    clients: String = null /* Default null, meaning all defined clients */ ,
-    multiProfile: Boolean = true): Directive1[AuthenticatedRequest] =
+               clients: String = null /* Default null, meaning all defined clients */ ,
+               multiProfile: Boolean = true,
+               enforceFormEncoding: Boolean = false, //Force form parameters to be passed for authentication or the request fails
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala]: Nice. I like this option best from all the ones we discussed

File: src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala [hunk 10]
```diff
@@ -43,16 +82,43 @@ class AkkaHttpSecurity[P <: CommonProfile](config: Config) {
...
-        }, actionAdapter, clients, "", "", multiProfile)
+        import ctx.materializer
+
+        /**
+          * First try to extract authentication credentials from form parameters using Akka's StrictForm unmarshallers.
+          * If that fails, either form encoding is enforced, in which case the request fails, or the request proceeds.
+          * If the request proceeds, other ways (e.g. basic auth) are assumed to be configured in pac4j in order to pass
+          * credentials.
+          */
+        Unmarshal(ctx.request.entity).to[StrictForm].fast.flatMap { form =>
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala]: Would it be possible to extract this into a separate (private) function

File: src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala [hunk 11]
```diff
@@ -43,16 +82,43 @@ class AkkaHttpSecurity[P <: CommonProfile](config: Config) {
...
+          val fields = form.fields.collect {
+            case (name, field) if name.nonEmpty =>
+              Unmarshal(field).to[String].map(fieldString => (name, fieldString))
+          }
+          Future.sequence(fields)
+        }.map { fields =>
+          AkkaHttpWebContext(ctx.request, fields)
+        }.recoverWith { case e =>
+          if(enforceFormEncoding) {
```
Comment (craffit) [inline src/main/scala/com/stackstate/pac4j/AkkaHttpSecurity.scala]: This recovers all errors happening during AkkaHttpWebContext construction, i'd move the recovery before the AkkaHttpWebContextConstruction. You can also turn the whole form field reading into a 'smart constructor' of AkkaHttpSecurity (one that takes just the HttpRequest and the enforceFormEncoding and returns a Future[AkkaHttpRequestContext])

File: src/test/scala/com/stackstate/pac4j/AkkaHttpSecurityTest.scala [hunk 12]
```diff
@@ -103,6 +104,107 @@ class AkkaHttpSecurityTest extends WordSpecLike with Matchers with ScalatestRout
...
+        "/",
+        entity = HttpEntity(ContentType(MediaTypes.`application/x-www-form-urlencoded`, HttpCharsets.`UTF-8`), "username=testuser".getBytes)
+      )
+
+      postRequest ~> akkaHttpSecurity.withAuthentication(enforceFormEncoding = true) { _ => complete("problem!") } ~> check {
+        status shouldEqual StatusCodes.OK
+      }
+    }
+
+    "fail a request when no parameters exist in a form and enforceFormEncoding is enabled" in {
```
Comment (craffit) [inline src/test/scala/com/stackstate/pac4j/AkkaHttpSecurityTest.scala]: Good coverage here
